server/schema_syncer.go

```diff
@@ -39,107 +39,113 @@ func (s *SchemaSyncer) Run(ctx context.Context, wg *sync.WaitGroup) {
-				}
-				instanceList, err := s.server.store.FindInstance(ctx, instanceFind)
-				if err != nil {
-					log.Error("Failed to retrieve instances", zap.Error(err))
-					return
-				}
-
-				for _, instance := range instanceList {
-					mu.Lock()
-					if _, ok := runningTasks[instance.ID]; ok {
-						mu.Unlock()
-						continue
-					}
-					runningTasks[instance.ID] = true
-					mu.Unlock()
-
-					go func(instance *api.Instance) {
-						log.Debug("Sync instance schema", zap.String("instance", instance.Name))
-						defer func() {
-							mu.Lock()
-							delete(runningTasks, instance.ID)
-							mu.Unlock()
-						}()
-						databaseList, err := s.server.syncInstance(ctx, instance)
-						if err != nil {
-							log.Debug("Failed to sync instance",
-								zap.Int("id", instance.ID),
-								zap.String("name", instance.Name),
-								zap.String("error", err.Error()))
-							return
-						}
-						for _, databaseName := range databaseList {
-							// If we fail to sync a particular database due to permission issue, we will continue to sync the rest of the databases.
-							if err := s.server.syncDatabaseSchema(ctx, instance, databaseName); err != nil {
-								log.Debug("Failed to sync database schema",
-									zap.Int("instanceID", instance.ID),
-									zap.String("instanceName", instance.Name),
-									zap.String("databaseName", databaseName),
-									zap.String("error", err.Error()))
-							}
-						}
-					}(instance)
-				}
-			}()
-		case instance := <-instanceSyncChan:
-			go func(instance *api.Instance) {
-				defer func() {
-					if r := recover(); r != nil {
-						err, ok := r.(error)
-						if !ok {
-							err = errors.Errorf("%v", r)
-						}
-						log.Error("Schema syncer PANIC RECOVER", zap.Error(err), zap.Stack("panic-stack"))
-					}
-				}()
-
-				databaseList, err := s.server.store.FindDatabase(ctx, &api.DatabaseFind{InstanceID: &instance.ID})
-				if err != nil {
-					log.Debug("Failed to find databases for the syncing instance",
-						zap.Int("id", instance.ID),
-						zap.String("name", instance.Name),
-						zap.String("error", err.Error()))
-					return
-				}
-				for _, database := range databaseList {
-					if database.SyncStatus != api.OK {
-						continue
-					}
-
-					// If we fail to sync a particular database due to permission issue, we will continue to sync the rest of the databases.
-					if err := s.server.syncDatabaseSchema(ctx, instance, database.Name); err != nil {
-						log.Debug("Failed to sync database schema",
-							zap.Int("instanceID", instance.ID),
-							zap.String("instanceName", instance.Name),
-							zap.String("databaseName", database.Name),
-							zap.String("error", err.Error()))
-					}
-				}
-			}(instance)
+			s.syncAllInstances(ctx)
+			// Sync all databases for all instances.
+			s.syncAllDatabases(ctx, nil /* instanceID */)
+		case instance := <-instanceDatabaseSyncChan:
+			// Sync all databases for instance.
+			s.syncAllDatabases(ctx, &instance.ID)
```

Comment:
We have synchronous s.syncInstance() before all the "instanceDatabaseSyncChan <- instance".

```
if _, err := s.syncInstance(ctx, updatedInstance); err != nil {
```

Reply:
Then, shall we move s.syncInstance here

Reply:
We need to keep the s.syncInstance() synchronous because we'd like to show up the list of databases for the instance once the instance is added/created. This also returns connection errors to users.